### ISSUE TITLE
Esri Feature Server

### ISSUE THREAD

**Issue by parthgundi** [TerriaJS/terriajs#4649]
Hi, i'm trying to add an esri feature server to my terria application but keep getting this error: TypeError: Cannot read property '0' of undefined
<img width="373" alt="Screenshot 2020-08-18 at 16 07 10" src="https://user-images.githubusercontent.com/30123754/90530503-e00eb800-e16c-11ea-88aa-343a3d54e7ee.png">

**Comment by rowanwins** [comment]
Hi @parthgundi 

Unfortunately it's hard to replicate based on that screenshot, can you please provide the URL you are using?

Thanks,
Rowan

**Comment by zoran995** [comment]
There was a discussion on gitter regarding this, terria removes token from URL and receive an error response from server ![error](https://user-content.gitter-static.net/def229321f22eadcea1de2e59cf2cd8bfcac668f/68747470733a2f2f66696c65732e6769747465722e696d2f3537303666613236313837626236663065616465363735322f6b466b7a2f7468756d622f696d6167652e706e67). 

According to that discussion url is [this feature server](https://services5.arcgis.com/2nCa3C5hWmWHagTM/ArcGIS/rest/services/thailand_property/FeatureServer/0/query?where=ObjectID%3E0&objectIds=&time=&geometry=&geometryType=esriGeometryEnvelope&inSR=&spatialRel=esriSpatialRelIntersects&resultType=none&distance=0.0&units=esriSRUnit_Meter&returnGeodetic=false&outFields=*&returnHiddenFields=false&returnGeometry=true&featureEncoding=esriDefault&multipatchOption=xyFootprint&maxAllowableOffset=&geometryPrecision=&outSR=&datumTransformation=&applyVCSProjection=false&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnExtentOnly=false&returnQueryGeometry=false&returnDistinctValues=false&cacheHint=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&having=&resultOffset=&resultRecordCount=&returnZ=false&returnM=false&returnExceededLimitFeatures=true&quantizationParameters=&sqlFormat=none&f=pgeojson&token=qt8pfpcgCUQSKGlsGgfPT_r34aePRSTFfOGYPnxY3rrt83Ybg-UeKmqAal3Q8gYHCVw70Kx5d-NI9Cqfs0UHyRwRL62W2kbI_yEFqPNHXxqqn_KMwLM56R_3h282Wi-Qp-mb7RP2d_swFiaq5JS7yBhIRxDSI1LDToOd8CmFpysMkikZN8T1dOD4hrsasPKGPOjB_FcNhofzm1GRBiLF4aXN4nvWqeqy5Z2jsHn9ZcM). I am thinking about list of allowed parameters for user added data, for example token shouldn't be removed. Also some better parsing of server response could be done so user get correct response, this error is thrown on [L329-L333](https://github.com/TerriaJS/terriajs/blob/master/lib/Models/ArcGisFeatureServerCatalogItem.js#L329-L333)

**Comment by rowanwins** [comment]
Esri tokens are generally short-lived so putting the token in the url probably won't work past 2 weeks (which I think is the longest you can generate a token for).

The typical way we handle this is via terriajs-server, see the [example here](https://github.com/TerriaJS/terriajs-server/blob/master/serverconfig.json.example#L68-L77), so that might provide an alternate avenue for investigation.

**Comment by zoran995** [comment]
That is a nice feature, maybe we could also add an option to catch access_token from URL and use it for loading. It will enable users to add their private resources from Esri feature and map servers. I am just not sure if there is an existing option that gives an indication if data is coming from `catalog` or  `User added data`.

User story is like this:
I add a few resources from the catalog and then realise it might be nice to also add information about buildings. That dataset only exists on my FeatureServer and it is private. I could generate an accessToken and add that data on the map viewer, currently, it will throw an error unless I make dataset public.

**Comment by rowanwins** [comment]
Yeah that's one option, although potentially a bit fiddily if people have to copy and paste their tokens around.

Another option that we've used in the [NSW Digital Twin](https://nsw.digitaltwin.terria.io/) is at the TerriaMap level we've added an Esri OAuth login button, this gives us a new token as required. Then in the index.js we make a patch to the `CorsProxy` class which appends the token when requesting the catalog item (even though we don't actually use the proxy to retrieve the data things still go via that class). If you'd like more details on this @zoran995 let me know and I can flick you some material via email.

**Comment by rowanwins** [comment]
And also of interest might be the newly added `ArcGisPortalGroup` item which can harvest items from something like ArcGIS Online or a hosted ArcGIS Portal instance

See
https://github.com/TerriaJS/terriajs/blob/next/lib/Traits/ArcGisPortalCatalogGroupTraits.ts
and 
https://github.com/TerriaJS/terriajs/blob/next/lib/Models/ArcGisPortalCatalogGroup.ts

Config looks like something
````
    {
      "type": "arcgis-portal-group",
      "name": "All open Content",
      "url": "https://portal.spatial.nsw.gov.au/portal",
      "groupBy": "portalCategories"
    }
````

**Comment by parthgundi** [comment]
Hi @rowanwins i would be interested in the esri OAuth login details that you have connected if possible thanks.
